papis/zenodo.py

```diff
@@ -0,0 +1,221 @@
+        "resource_type", [{"key": "type", "action": lambda x: x["type"]}]
+    ),
+    KeyConversionPair("revision", [{"key": "revision", "action": None}]),
+    KeyConversionPair("license", [{"key": "license", "action": lambda x: x["id"]}]),
+    KeyConversionPair(
+        "description", [{"key": "description", "action": get_text_from_html}]
+    ),
+]
+
+
+def zenodo_data_to_papis_data(data: Dict[str, Any]) -> Dict[str, Any]:
+    """Converts the dictionary from Zenodo to the conventional papis format.
+
+    :param data: the raw dictionary from Zenodo
+    :type data: Dict[str, Any]
+    :return: a new dictionary in a conventional papis form
+    :rtype: Dict[str, Any]
+    """
+    # Merge metadata into data
+    data.update(data["metadata"])
+    del data["metadata"]
+
+    new_data = papis.document.keyconversion_to_data(key_conversion, data)
+    new_data["file_urls"] = data["files"]
+
+    return new_data
+
+
+def is_valid_record_id(record_id: str) -> bool:
+    """Checks if a record is a valid Zenodo record, first checking its form and then
+    testing against Zenodo
+
+    :param record_id: a Zenodo record id
+    :type record_id: str
+    :return: whether the record is valid
+    :rtype: bool
+    """
+    record_id = record_id.strip()
+    if not record_id.isdigit():
+        return False
+
+    with papis.utils.get_session() as session:
+        response = session.get(ZENODO_URL.format(record_id=record_id))
+
+    return response.ok
+
+
+def get_data(record_id: str) -> Dict[str, Any]:
+    """Fetches a record from the Zenodo API and processes it with a helper function
+
+    :param record_id: a Zenodo record id
+    :type record_id: str
+    :return: a processed zenodo record
+    :rtype: Dict[str, Any]
+    """
+    with papis.utils.get_session() as session:
+        response = session.get(
+            ZENODO_URL.format(record_id=record_id.strip()),
+            headers={"user-agent": f"papis/{papis.__version__}"},
+        )
+
+    try:
+        json_data = json.loads(response.content.decode())
+    except json.JSONDecodeError as exc:
+        logger.error("Failed to decode response from Zenodo.", exc_info=exc)
+
+    return zenodo_data_to_papis_data(json_data)
+
+
+class LinkInfo(TypedDict):
+    """Represents a the url for a Zenodo record.
+    Attributes:
+    - id, the uuid of the file
+    """
+
+    self: str
+
+
+class FileInfo(TypedDict):
+    """Represents a file info entry from a Zenodo record.
+
+    - id, the uuid of the file
+    - key, the filename
+    - size: the size of a file in bytes
+    - checksum: md5sum hash of the file.
+    - links: dict with the url to download the file
+    """
+
+    id: str
+    key: str
+    size: int
+    checksum: str
+    links: LinkInfo
+
+
+class Importer(papis.importer.Importer):
+    """Importer downloading data from a Zenodo ID"""
+
+    def __init__(self, uri: str = "") -> None:
+        super().__init__(name="zenodo", uri=uri)
+        self._file_urls: List[FileInfo] = []
```

Comment:
Hm, this is a bit ugly. We should probably improve the API for these importers to let them cache things in a nicer way..